### DISCUSSION TITLE
How to set custom font

### DISCUSSION THREAD

**Discussion by TrionProg** [emilk/egui#1420] [Q&A]
Hello. I want to set custom font for, for example, button of special type.

As I understand, I need add FontData to fonts, this font is part of font family, for example, bold font

`let mut fonts = egui::FontDefinitions::default();
 fonts.font_data.insert(
        "Helvetica".to_owned(),
        egui::FontData::from_static(include_bytes!("../../assets/LinLibertine_R_G.ttf"))
);`

I need add this special version of font to FontFamily, I need create it. But I can not understand, what Vec<String> means and priority or how select italic/bold/mono versions? I think, String must be name of FontData
` fonts.families.insert(egui::FontFamily::Name("Helvetica".into()), vec!["Helvetica".to_owned()]);`

And set font
`ctx.set_fonts(fonts);`

Next, I want add font Helvetica with size 24 to style:
`    let mut style = Style::default();

    let font_id = FontId {
        size: 24.0,
        family: FontFamily::Name("Helvetica".into())
    };

    style.text_styles.insert(TextStyle::Name("ButtonFont".into()), font_id);
    
    ctx.set_style(Arc::new(style));
    `
And, next, I am using font:
`let text = text.into_galley(ui, wrap, wrap_width, TextStyle::Name("ButtonFont".into()));`

But I have error `FontFamily::{:?} is not bound to any fonts`

**Comment by emilk** [comment]
Are you sure you call `set_fonts` before using the new font? What is the full `FontDefinitions/set_fonts` code look like?

**Reply by TrionProg** [reply]
method of Application
```
fn update(&mut self, ctx: &egui::Context, frame: &epi::Frame) {
        let Self { label, value, menu } = self;

        crate::gui::assets::load_assets(ctx); //assets is static struct, it loades once if not loaded
        let assets = crate::gui::assets::get_assets();
        ctx.set_style(assets.style.clone());
        

        let mut fonts = egui::FontDefinitions::default();
        fonts.font_data.insert("Helvetica".to_owned(),
                               egui::FontData::from_static(include_bytes!("../../assets/LinLibertine_R_G.ttf")));

        fonts.families.insert(egui::FontFamily::Name("Helvetica".into()), vec!["Helvetica".to_owned()]);
        
        fonts.families.get_mut(&egui::FontFamily::Proportional).unwrap() //it works
            .insert(0, "Helvetica".to_owned());

        fonts.families.get_mut(&egui::FontFamily::Monospace).unwrap()
            .insert(0, "Helvetica".to_owned());//.push("Helvetica".to_owned());


        ctx.set_fonts(fonts);
        
        //using font...
        egui::CentralPanel::default().show(ctx, |ui| { ... }
        ```
        
Creation of styles, once and before font loading(.set_font)
  ``` 
  pub fn load_assets(context: &egui::Context) {
  
  let (style, fonts) = create_style();
  
  *assets = Some(new_assets);
  }

fn create_style() -> (Arc<Style>, Fonts) {
    use egui::{FontId, FontFamily, TextStyle};

    let mut style = Style::default();

    let button_monumental:Arc<str> = "Helvetica".into();

    let font_id = FontId {
        size: 24.0,
        family: FontFamily::Name("Helvetica".into())
        //family: FontFamily::Name("Linux Libertine G".into())
    };

    style.text_styles.insert(TextStyle::Name(button_monumental.clone()), font_id);

    for (t, t2) in style.text_styles.iter() {
        println!("{:?} {:?}", t, t2.family);

    }

    let fonts = Fonts {
        button_monumental
    };

    (Arc::new(style), fonts)
}
```

**Reply by emilk** [reply]
You are installing the font's every frame, which is a really bad idea. Set up the fonts ONCE in `App::setup` instead: https://github.com/emilk/egui/blob/0.17.0/eframe/examples/custom_font.rs

**Reply by TrionProg** [reply]
Hello. I've not expected, but after moving code above into App::setup after

`if let Some(storage) = _storage {
            *self = epi::get_value(storage, epi::APP_KEY).unwrap_or_default()
        }`
        
It works! Also in setup are loading of images and other content. But I am scared, I can not load fonts in runtime, for example, images, then they are needed, to make applocation loading faster and do not show dead window on application launch. TextAlign to center for buttons I made by calculating button_x + (button_width - galley.rect.(max.x-min.x))/2 . Thank you!

And I can not understand logic:
Here you can add sub-families of fonts. But how egui selects font to use?
`fonts.families.insert(egui::FontFamily::Name("Helvetica".into()), vec!["Helvetica".to_owned()]);`

For example, font-family is Helvetica, but Bold, Italic must be in this vector?
